kafka-1.0.x/src/main/scala/monix/kafka/CommittableOffsetBatch.scala

```diff
@@ -0,0 +1,82 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package monix.kafka
+
+import monix.eval.Task
+import org.apache.kafka.common.TopicPartition
+
+/** Batch of Kafka offsets which can be committed together.
+  * Can be builded from offsets sequence by [[CommittableOffsetBatch#apply]] method.
+  * Besides you can use [[CommittableOffsetBatch#empty]] method to create empty batch and
+  * add offsets to it by [[updated]] method.
+  *
+  * Anyway offsets order make sense! Only last added offset for topic and partition will
+  * be committed to Kafka.
+  *
+  * @param offsets is the offsets batch for a few topics and partitions.
+  *        Make sure that each of them was received from one [[KafkaConsumerObservable]].
+  *
+  * @param commitBatch is the function for batched commit realized as closure
+  *        in [[KafkaConsumerObservable]] context. This decision was made for
+  *        thread-safety reasons. This parameter is obtained from last [[CommittableOffset]]
+  *        added to batch.
+  */
+final class CommittableOffsetBatch(
+  val offsets: Map[TopicPartition, Long],
+  commitBatch: Map[TopicPartition, Long] => Task[Unit]) {
+
+  /**
+    * Commits [[offsets]] to Kafka
+    * */
+  def commit(): Task[Unit] = commitBatch(offsets)
+
+  /**
+    * Adds new [[CommittableOffset]] to batch. Added offset replaces previous one specified
+    * for same topic and partition.
+    * */
+  def updated(committableOffset: CommittableOffset): CommittableOffsetBatch =
+    new CommittableOffsetBatch(
+      offsets.updated(committableOffset.topicPartition, committableOffset.offset),
+      committableOffset.commitBatch
+    )
+}
+
+object CommittableOffsetBatch {
+
+  /**
+    * Creates empty [[CommittableOffsetBatch]]. Can be used as neutral element in fold:
+    * {{{
+    *   offsets.foldLeft(CommittableOffsetBatch.empty)(_ updated _)
+    * }}}
+    * */
+  def empty: CommittableOffsetBatch = new CommittableOffsetBatch(Map.empty, _ => Task.pure(()))
+
+  /**
+    * Builds [[CommittableOffsetBatch]] from offsets sequence. Be careful with
+    * sequence order. If there is more than once offset for a topic and partition in the
+    * sequence then the last one will remain.
+    * */
+  def apply(offsets: Seq[CommittableOffset]): CommittableOffsetBatch =
+    if (offsets.nonEmpty) {
+      val aggregatedOffsets = offsets.foldLeft(Map.empty[TopicPartition, Long]) { (acc, o) =>
+        acc.updated(o.topicPartition, o.offset)
+      }
+      new CommittableOffsetBatch(aggregatedOffsets, offsets.head.commitBatch)
```

Comment:
I think this might be dangerous if users can create their own `CommittableOffset`s. 
Also, do you think it would be possible to support merging different `Observables`, each from different topic/group ?

Reply:
> I think this might be dangerous if users can create their own CommittableOffsets.

Agreed. I make `CommittableOffsetBatch`'s and `CommittableOffset`'s constructors `private[kafka]`.

> Also, do you think it would be possible to support merging different Observables, each from different topic/group ?

Hmmm, sounds very dangerous. We can store a few `commitBatch` callbacks from different consumers and carefully commit batches corresponding to them. But I do not understand how much it is necessary and tend to be against this possibility.

Reply:
> > I think this might be dangerous if users can create their own CommittableOffsets.
> 
> Agreed. I make `CommittableOffsetBatch`'s and `CommittableOffset`'s constructors `private[kafka]`.
> 

It's possible to create `CommittableOffset` using its `apply` method - is it intended? I guess it could be fine but if someone does it then he probably wants to make sure that his commit function will be used

> > Also, do you think it would be possible to support merging different Observables, each from different topic/group ?
> 
> Hmmm, sounds very dangerous. We can store a few `commitBatch` callbacks from different consumers and carefully commit batches corresponding to them. But I do not understand how much it is necessary and tend to be against this possibility.

Alright, I don't think it's that important

Reply:
> It's possible to create `CommittableOffset` using its `apply` method - is it intended?

No longer possible :sweat_smile: